src/ray/core_worker/experimental_mutable_object_manager.cc

```diff
@@ -159,230 +172,191 @@ void ExperimentalMutableObjectManager::DestroySemaphores(const ObjectID &object_
+  data = SharedMemoryBuffer::Slice(object->buffer, 0, data_size);
+  if (metadata) {
     // Copy the metadata to the buffer.
-    memcpy((*data)->Data() + channel.mutable_object->header->data_size,
-           metadata,
-           metadata_size);
+    memcpy(data->Data() + data_size, metadata, metadata_size);
   }
+  channel->written = true;
   return Status::OK();
 }
 
-Status ExperimentalMutableObjectManager::WriteRelease(const ObjectID &object_id) {
-  auto writer_channel_entry = writer_channels_.find(object_id);
-  if (writer_channel_entry == writer_channels_.end()) {
+Status MutableObjectManager::WriteRelease(const ObjectID &object_id) {
+  WriterChannel *channel = GetChannel<WriterChannel>(writer_channels_, object_id);
+  if (!channel) {
     return Status::ObjectNotFound("Writer channel has not been registered");
   }
-
-  auto &channel = writer_channel_entry->second;
-  if (channel.is_sealed) {
-    return Status::Invalid("Must WriteAcquire before WriteRelease on a mutable object");
-  }
+  RAY_CHECK(channel->written);
 
   PlasmaObjectHeader::Semaphores sem = GetSemaphores(object_id);
-  RAY_RETURN_NOT_OK(channel.mutable_object->header->WriteRelease(sem));
-  channel.is_sealed = true;
-
-  return Status::OK();
-}
-
-Status ExperimentalMutableObjectManager::RegisterReaderChannel(
-    const ObjectID &object_id, std::unique_ptr<plasma::MutableObject> mutable_object) {
-  auto inserted =
-      reader_channels_.emplace(object_id, ReaderChannel(std::move(mutable_object)));
-  if (!inserted.second) {
-    return Status::Invalid("Reader channel already registered");
-  }
-  OpenSemaphores(object_id);
-  RAY_LOG(DEBUG) << "Registered reader channel " << object_id;
+  std::unique_ptr<plasma::MutableObject> &object = channel->mutable_object;
+  RAY_RETURN_NOT_OK(object->header->WriteRelease(sem));
+  channel->written = false;
   return Status::OK();
 }
 
-bool ExperimentalMutableObjectManager::ReaderChannelRegistered(
-    const ObjectID &object_id) const {
-  return reader_channels_.count(object_id);
-}
-
-Status ExperimentalMutableObjectManager::ReadAcquire(const ObjectID &object_id,
-                                                     std::shared_ptr<RayObject> *result)
+Status MutableObjectManager::ReadAcquire(const ObjectID &object_id,
+                                         std::shared_ptr<RayObject> &result)
     ABSL_NO_THREAD_SAFETY_ANALYSIS {
-  RAY_LOG(DEBUG) << "ReadAcquire  " << object_id;
-  auto reader_channel_entry = reader_channels_.find(object_id);
-  if (reader_channel_entry == reader_channels_.end()) {
+  ReaderChannel *channel = GetChannel<ReaderChannel>(reader_channels_, object_id);
+  if (!channel) {
     return Status::ObjectNotFound("Reader channel has not been registered");
   }
-  auto &channel = reader_channel_entry->second;
 
-  // This lock ensures that there is only one reader at a time. The lock is released in
-  // `ReadRelease()`.
-  channel.lock->Lock();
+  channel->lock->Lock();
```

Comment:
Could you add back the comment that was here?

Reply:
Done.